src/agent/coverage/src/disasm.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::collections::BTreeSet;
+
+use anyhow::Result;
+use iced_x86::{Decoder, DecoderOptions, Instruction};
+
+use crate::code::{ModuleIndex, Symbol};
+
+pub struct ModuleDisassembler<'a> {
+    module: &'a ModuleIndex,
+    data: &'a [u8],
+}
+
+impl<'a> ModuleDisassembler<'a> {
+    pub fn new(module: &'a ModuleIndex, data: &'a [u8]) -> Result<Self> {
+        Ok(Self { module, data })
+    }
+
+    /// Find block entry points for every symbol in the module.
+    pub fn find_blocks(&self) -> BTreeSet<u64> {
+        let mut blocks = BTreeSet::new();
+
+        for symbol in self.module.symbols.iter() {
+            let sym_blocks = self.find_symbol_blocks(symbol);
+            blocks.extend(sym_blocks);
+        }
+
+        blocks
+    }
+
+    /// Find all entry points for blocks contained within the region of `symbol`.
+    pub fn find_symbol_blocks(&self, symbol: &Symbol) -> BTreeSet<u64> {
+        // Slice the symbol's instruction data from the module file data.
+        let data = {
+            let lo = symbol.file_offset as usize;
+            let hi = lo + (symbol.size as usize);
+            &self.data[lo..hi]
+        };
+
+        // Initialize a decoder for the current symbol.
+        let mut decoder = Decoder::new(64, data, DecoderOptions::NONE);
+
+        // Compute the VA of the symbol, assuming preferred module base VA.
+        let va = self.module.base_va + symbol.image_offset;
```

Comment:
Do we need a checked add?

Reply:
Yes, that'd be good to do here, since we can't locally validate that in the `Symbol` ctor.